account-keys/src/identity.rs

```diff
@@ -0,0 +1,198 @@
+// Copyright (c) 2018-2020 MobileCoin Inc.
+
+//! In cryptography, several private keys can be derived from a single source of
+//! entropy using a strong KDF (key derivation function).
+//! This is sound, so long as the input-key-material to the KDF itself has
+//! at least enough entropy as the length of any one of the derived keys.
+//!
+//! The RootIdentity object contains 32 bytes of "root entropy", used with HKDF
+//! to produce the other MobileCoin private keys. This is useful because an AccountKey
+//! derived this way can be represented with a smaller amount of information.
+//!
+//! The other (fog-related) fields of RootIdentity are analogous to AccountKey.
+
+use crate::{AccountKey, RootEntropyProblem};
+use alloc::{
+    string::{String, ToString},
+    vec::Vec,
+};
+use core::{convert::From, hash::Hash};
+use curve25519_dalek::scalar::Scalar;
+use hkdf::Hkdf;
+use mc_crypto_hashes::Blake2b256;
+use mc_crypto_keys::RistrettoPrivate;
+use prost::Message;
+use rand_core::{CryptoRng, RngCore};
+use serde::{Deserialize, Serialize};
+
+pub const TEST_FOG_AUTHORITY_FINGERPRINT: [u8; 4] = [9, 9, 9, 9];
+pub const TEST_FOG_REPORT_KEY: &str = "";
+
+/// A RootIdentity contains 32 bytes of root entropy (for deriving private keys
+/// using a KDF), together with any fog data for the account.
+#[derive(Clone, PartialEq, Eq, Hash, Message, Serialize, Deserialize)]
+pub struct RootIdentity {
+    /// Root entropy used to derive a user's private keys.
+    #[prost(bytes, tag = 1)]
+    pub root_entropy: Vec<u8>,
+    /// Fog report url
+    #[prost(string, tag = 2)]
+    pub fog_url: String,
+    /// Fog report id
+    #[prost(string, tag = 3)]
+    pub fog_report_id: String,
+    /// Fog authority fingerprint
+    #[prost(bytes, tag = 4)]
+    pub fog_authority_fingerprint: Vec<u8>,
+}
+
+impl RootIdentity {
+    /// Generate a random root identity with a specific fog url configured
+    pub fn random<T: RngCore + CryptoRng>(rng: &mut T, fog_url: Option<&str>) -> Self {
+        let mut root_entropy = [0u8; 32];
+        // Filter out bad root entropies
+        // TODO: Break after 100 tries and fail?
```

Comment:
i feel like, maybe a better way is, use a known good rng with fixed seed, and confirm in a unit test that it succeeds within at least 100 tries. if a developer is using this code and their thing is in an infinite loop, it implies that something is wrong with their Rng i guess.

Reply:
As noted above, I think it's better to just run some tests on the RNG at startup, then just use it normally everywhere else. Though TBH, `CryptoRng` is supposed to be that already...